### PR TITLE
Via IR bytecode comparison

### LINKED ISSUE #13583: Bytecode compare runs with `--via-ir`
The IR pipeline is non-experimental now but our [bytecode compare job](https://github.com/ethereum/solidity/blob/v0.8.17/.circleci/config.yml#L1469-L1504) still uses the legacy codegen. We should add extra runs with `--via-ir`.

I think it's enough to do this for CLI. We might also switch the JSON runs to `--via-ir` but I don't think it's necessary to run them for both codegens.

Note that solc-bin has a [bytecode compare PR check](https://github.com/ethereum/solc-bin/blob/gh-pages/.github/workflows/t-bytecode-compare.yml) that's implemented separately. Not sure if we want to update that one. Running it with one codegen already provides most of its value (which is making sure that the binary is not damaged). The binaries submitted to solc-bin are expected to have gone through the full bytecode comparison in the main repo anyway.

### PR DESCRIPTION
**Pull request by cameel**
~Depends on #14354.~ Merged.
Depends on #14374.
Fixes #13583.

This is the final PR for `--via-ir` bytecode comparison.

### CODE REVIEW

File: .circleci/config.yml
Comment (cameel) [inline .circleci/config.yml:91]: Unfortunately even after all that upfront work to get things working in parallel, it's pretty slow. Optimized via IR preset takes as long as 25 min on windows and on emscripten.

**Comment by cameel** [comment]
Timing, from the [most recent run](https://app.circleci.com/pipelines/github/ethereum/solidity/30336/workflows/5a907609-222a-4e5a-aa40-13ec169ac1f5):

| Job                     | Preset         | Legacy | Via IR  |
|-------------------------|----------------|--------|---------|
| `b_bytecode_win`        | `optimize`     | 6m 52s | 24m 24s |
| `b_bytecode_win`        | `no-optimize`  | 5m 5s  | 10m 12s |
| `b_bytecode_osx`        | `optimize`     | 3m 56s | 10m 42s |
| `b_bytecode_osx`        | `no-optimize`  | 3m 38s | 5m 55s  |
| `b_bytecode_ems`        | `optimize`     | 5m 31s | 26m 47s |
| `b_bytecode_ems`        | `no-optimize`  | 2m 8s  | 13m 25s |
| `b_bytecode_ubu`        | `optimize`     | 3m 48s | 9m 18s  |
| `b_bytecode_ubu`        | `no-optimize`  | 3m 2s  | 5m 9s   |
| `b_bytecode_ubu_static` | `optimize`     | 2m 12s | 7m 51s  |
| `b_bytecode_ubu_static` | `no-optimize`  | 1m 13s | 3m 35s  |

**Comment by cameel** [comment]
Also, we have a problem: [`t_bytecode_compare`](https://app.circleci.com/pipelines/github/ethereum/solidity/30336/workflows/5a907609-222a-4e5a-aa40-13ec169ac1f5/jobs/1346462) does not pass via IR. And it looks like it's due to actual differences in generated bytecode. Weirdly, this happens only for unoptimized code. Optimized is fine.

I will investigate that on Monday.

**Comment by cameel** [comment]
Current status: blocked by #14361, which I'll be fixing.

If need be, I could add a temporary workaround to delete test cases that do not pass and merge this. I would then remove that workaround in the PR fixing #14361.

**Comment by cameel** [comment]
With #14374 and one additional tweak the bytecode comparison finally passes.

The tweak is basically that one of the tests triggers a `YulException` (actually stack too deep, but for some reason it's not `StackTooDeep`) and CLI just barfs while Standard JSON can still successfully return some outputs like metadata. This is basically a hack to ignore those outputs. The proper way to deal with it would be to fix #13925.

The PR is now reviewable but I'll keep it as a draft until #14374 is merged because PRs with dependencies always confuse people :)

**Review by matheusaaguiar — DISMISSED** [review]
LGTM

**Comment by cameel** [comment]
Base PR was merged. This is now reviewable.

@matheusaaguiar But since the PR is simple and you already reviewed and approved it, we can also just merge if you reapprove.

File: .circleci/parallel_bytecode_report.sh
Comment (cameel) [inline .circleci/parallel_bytecode_report.sh:48]: @ekpyron dropped #13496 from the focus board recently so I guess he does not want use to focus on fixing in this. For that reason it'll stay like this for the time being. I.e. we're not really running the comparison on **all** test cases from the repo.

File: .circleci/config.yml
Comment (nikola-matic) [reply .circleci/config.yml:91]: What does `no_output_timeout` do? Fail the step if the runtime exceeds 30 minutes? If so, then a 5 minute buffer for a job that takes 25 minutes is a little tight, no? Would it make sense to bump this a little bit; maybe 35 minutes?

**Review by nikola-matic — APPROVED** [review]

File: .circleci/config.yml
Comment (cameel) [reply .circleci/config.yml:91]: Yeah. But the default timeout is 10 min so I'm increasing it, not decreasing it :)

I think 30 min is fine. If it fails we can still increase it but we'll get a clear signal that things are getting out of hand. 25 min is actually way more than I expected. I'd prefer this to finish much sooner. If not for the fact that we're working on IR performance already, I'd suggest parallelizing things even more to improve it, rather than keep bumping the limit.

**Pull request merged**
